### DISCUSSION TITLE
Adding QED symbol at the end of an exercise's solution?

### DISCUSSION THREAD

**Discussion by darthlite** [quarto-dev/quarto-cli#7314] [Q&A]
### Description

How can I add a QED symbol by default at the end of every exercise's solution just to clearly indicate that the exercise is done.

I use:

::: .solution
Solution here.
:::

**Comment by mcanouil** [comment]
Which format?

Could you share a small self-contained "working" (reproducible) example to work with, _i.e._, a complete Quarto document or a Git repository? Thanks.

You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Comment by cscheid** [comment]
That's not exactly simple to do right now because solution nodes are not easily targetable in the current version of Quarto. I hope to land support for that next week, as part of #4944 .

**Reply by darthlite** [reply]
Thanks for the quick reply! I see. Is this part of v1.4? If so, is there an ETA for 1.4?

**Comment by cscheid** [comment]
> Is this part of v1.4?

Yes.

> If so, is there an ETA for 1.4?

We hope to get it out by the end of the year.

**Comment by darthlite** [comment]
Hi, just checking in to see if this can be done now?

Thanks!

**Reply by cscheid** [reply]
The answer is a qualified "yes". I realize now that I implemented this somewhat badly and I will have to fix it in the future. So this code will unfortunately break in the future (it'll be an easy fix, but it'll break).

Minimal example:


## test.qmd

````qmd
---
format: html
filters:
  - addqedsymbol.lua
---

::: solution

Solution here.

:::
````

## addqedsymbol.lua

````lua
function Proof(proof)
    if proof.type ~= "Solution" then
        return nil
    end
    local content = proof.div.content
    local last_para = content[#content]
    if last_para.t ~= "Para" then
        print("Error: proof block does not end with a paragraph. Ignoring.")
        return nil
    end
    last_para.content:insert(pandoc.Str(" □"))
    return quarto.Proof(proof)
end
````

This doesn't look super great - the symbol is not aligned to the right, this won't work in LaTeX, etc. But it should be enough to get you started.

**Reply by darthlite** [reply]
Thanks a bunch!

I managed to tweak this a tiny bit and make it work with pdf/latex and produce the symbol as desired.

Any heads up as to what form the future breakage will take (files won't render, symbol won't appear when this breaks, or something else?) just to try and future-proof this a bit if I'm able to and/or add comments for future me?

**Reply by cscheid** [reply]
The redesign that's needed is to change `proof.div.content` to `proof.content`. We're exposing `div` as a field but that should really not be done; we should have meaningful field names (like FloatRefTargets have with `content` and `caption_long`). I can't predict the error message, but the necessary change will be pretty small.
